Domain/SecurityIdentityRetrievalStrategy.php

```diff
@@ -39,37 +39,54 @@ public function __construct(RoleHierarchyInterface $roleHierarchy, Authenticatio
 
     /**
      * {@inheritdoc}
+     *
+     * @return RoleSecurityIdentity[]
      */
-    public function getSecurityIdentities(TokenInterface $token)
+    public function getSecurityIdentities(?TokenInterface $token)
     {
         $sids = [];
 
         // add user security identity
-        if (!$token instanceof AnonymousToken) {
+        if (!$token instanceof AnonymousToken && null !== $token) {
             try {
                 $sids[] = UserSecurityIdentity::fromToken($token);
             } catch (\InvalidArgumentException $e) {
                 // ignore, user has no user security identity
             }
         }
 
-        // add all reachable roles
-        foreach ($this->roleHierarchy->getReachableRoleNames($token->getRoleNames()) as $role) {
-            $sids[] = new RoleSecurityIdentity($role);
+        if (null !== $token) {
+            // add all reachable roles
+            foreach ($this->roleHierarchy->getReachableRoleNames($token->getRoleNames()) as $role) {
+                $sids[] = new RoleSecurityIdentity($role);
+            }
         }
 
+        $anonymousRole = \defined('\Symfony\Component\Security\Core\Authorization\Voter\AuthenticatedVoter::PUBLIC_ACCESS') ? AuthenticatedVoter::PUBLIC_ACCESS : AuthenticatedVoter::IS_AUTHENTICATED_ANONYMOUSLY;
         // add built-in special roles
         if ($this->authenticationTrustResolver->isFullFledged($token)) {
             $sids[] = new RoleSecurityIdentity(AuthenticatedVoter::IS_AUTHENTICATED_FULLY);
             $sids[] = new RoleSecurityIdentity(AuthenticatedVoter::IS_AUTHENTICATED_REMEMBERED);
-            $sids[] = new RoleSecurityIdentity(AuthenticatedVoter::IS_AUTHENTICATED_ANONYMOUSLY);
+            $sids[] = new RoleSecurityIdentity($anonymousRole);
         } elseif ($this->authenticationTrustResolver->isRememberMe($token)) {
             $sids[] = new RoleSecurityIdentity(AuthenticatedVoter::IS_AUTHENTICATED_REMEMBERED);
-            $sids[] = new RoleSecurityIdentity(AuthenticatedVoter::IS_AUTHENTICATED_ANONYMOUSLY);
-        } elseif ($this->authenticationTrustResolver->isAnonymous($token)) {
-            $sids[] = new RoleSecurityIdentity(AuthenticatedVoter::IS_AUTHENTICATED_ANONYMOUSLY);
+            $sids[] = new RoleSecurityIdentity($anonymousRole);
+        } elseif ($this->isNotAuthenticated($token)) {
+            $sids[] = new RoleSecurityIdentity($anonymousRole);
         }
```

Comment:
I've never used ACL, so I'm not sure about this, but I guess change the security identity is a BC break (as these are persisted, so `IS_AUTHENTICATED_ANONYMOUSLY` might be persisted still?). Can you confirm (or reject) this?

If this is the case, we should support both token names I think (with hardcoded strings) and then somehow trigger a deprecation when the anonymous role is used.

Reply:
Hmm, you're right. I will add the old role back, but I'm not sure on how to deprecate this? Would be a warning/notice in an upgrade file be enough? Or even the symfony deprecation itself is enough?

Reply:
It would be great if we can add a specialized deprecation notice about it.

Can we do something in `AclProvider::findAcls()`, that triggers whenever we fetch a security identity with the old attribute name?